utils/guard/guard.go

```diff
@@ -0,0 +1,22 @@
+package guard
+
+import "sync"
+
+// Guard is a thread-safe wrapper around T.
+// It guarantees a thread-safe access to the underlying value T.
```

Comment:
can you also comment on the recommended use of this struct?
if a struct has 10 members that require thread-safety, do you recommend using this for every member?

Reply:
It depends on the particular situation. Generally, fields that must be kept "in sync" must be protected with a single lock to avoid an invalid intermediate state. Otherwise, each field could be put in a separate `Guard`.

Reply:
right. i am a little wary that future devs will use guard for fields that are always read/updated together, hence the request to add comments for guidance.

Reply:
also, what is your consideration for not choosing atomic.Value here?
for me, it seems a more natural choice if the only purpose is a thread-safe update for a single variable (and not semaphoring a sequence of events).

Reply:
In case you want type-safety there is also https://pkg.go.dev/sync/atomic#Pointer that allows atomic access to a value.

Reply:
> also, what is your consideration for not choosing atomic.Value here? for me, it seems a more natural choice if the only purpose is a thread-safe update for a single variable (and not semaphoring a sequence of events).

The reason for choosing this `Guard` over the previous atomic pointer is simplicity and safety. It doesn't require unsafe and pointer casting.

The reason for choosing `Guard` over just sticking a mutex next to the guarded variable into a struct is readability and safety (at the cost of the simplicity of use):
- ➕  readability: it's clear which variable is guarded. When a mutex is just stuck in a struct, it is not clear what it is supposed to protect
- ➕ safety: it's impossible to access the guarded variable without a lock
- ➖ lost simplicity: it might not be obvious/easy to use this `Apply` API

@fasmat I didn't know that generic pointer, thx! It removes my main concerns regarding typecasting and using unsafe. I'm OK with using it for the pending poet clients.

Reply:
> I'm OK with using it for the pending poet clients.

please do. thanks. this solution is perfect. thank you @fasmat 

my rationale for singling this out in this PR (i have not reviewed the rest)
- the Apply() with func ptr smells over-engineered for simply providing thread-safety to a variable.
- it's all too easy to start using Guard for every variable for a new programmer, which will cause a lot of unnecessary locking ops.
- the lib also ignored the variable's use characteristic (read/write frequency)

either of the following approach is good for me:
- use a generic library Guard but with solutions from atomic pkg and a simple getter/setter
- use a sync.Mutex next to the variable and use simple getter/setter

Reply:
@fasmat Rewrote to the generic pointer.